src/intervals/intervalstream.jl

```diff
@@ -0,0 +1,180 @@
+end
+
+
+# Do we need a state if this is going to be stateful. State can maybe just
+# be the current interval in b
+function Base.next{S, T, U, V}(it::IntervalStreamIntersectIterator{S, T},
+                               state::IntervalStreamIntersectIteratorState{U, V})
+    value = (it.a_buffer[state.a_buffer_pos], it.b_interval)
+    a_state = state.a_state
+    b_state = state.b_state
+
+    # another intersection in a_buffer?
+    a_buffer_pos = state.a_buffer_pos + 1
+    while a_buffer_pos <= length(it.a_buffer) &&
+          !isoverlapping(it.a_buffer[a_buffer_pos], it.b_interval)
+        a_buffer_pos += 1
+    end
+
+    # look for an intersection for b_interval by reading more entries from a
+    b_interval = it.b_interval
+    if a_buffer_pos > length(it.a_buffer) && !done(it.a, a_state)
+        while true
+            a_interval, a_state = next(it.a, a_state)
+            if isless(a_interval, it.a_buffer[end], it.seqname_isless)
+                error("Interval streams must be sorted to perform intersection.")
+            end
+            push!(it.a_buffer, a_interval)
+
+            if isoverlapping(b_interval, it.a_buffer[end]) ||
+               precedes(b_interval, it.a_buffer[end]) ||
+               done(it.a, a_state)
+                break
+            end
+            a_buffer_pos += 1
+        end
+
+        if !isoverlapping(it.a_buffer[a_buffer_pos], b_interval)
+            a_buffer_pos += 1
+        end
+    end
+
+    # no more intersections against b_interval, try next interval from b
+    while (a_buffer_pos > length(it.a_buffer) ||
+           !isoverlapping(it.a_buffer[a_buffer_pos], b_interval)) && !done(it.b, b_state)
+
+        b_interval_next, b_state = next(it.b, b_state)
+        if isless(b_interval_next, b_interval, it.seqname_isless)
+            error("Interval streams must be sorted to perform intersection.")
+        end
+        b_interval = b_interval_next
+
+        # evict as much as possible from a_buffer
+        a_interval = it.a_buffer[1]
+        while !isempty(it.a_buffer) && precedes(it.a_buffer[1], b_interval, it.seqname_isless)
+            a_interval = shift!(it.a_buffer)
+        end
+
+        if isempty(it.a_buffer)
+            a_interval, b_interval, a_state, b_state = first_intersection(
+                it.a, it.b, a_state, b_state, a_interval, b_interval, it.seqname_isless)
+            it.b_interval = b_interval
+            if isoverlapping(a_interval, b_interval)
+                push!(it.a_buffer, a_interval)
+                a_buffer_pos = 1
+            end
+            break
+        elseif isoverlapping(it.a_buffer[1], b_interval)
+            a_buffer_pos = 1
+            it.b_interval = b_interval
+        end
+    end
+
+    return value, IntervalStreamIntersectIteratorState{U, V}(
+                        a_buffer_pos, a_state, b_state)
+end
+
+
+function Base.done{S, T, U, V}(it::IntervalStreamIntersectIterator{S, T},
```

Comment:
This generic function is imported explicit on this module, therefore we don't need `Base.` here.